library/x509_crt.c

```diff
@@ -1403,10 +1404,39 @@ int mbedtls_x509_crt_parse_path( mbedtls_x509_crt *chain, const char *path )
             ret = MBEDTLS_ERR_X509_BUFFER_TOO_SMALL;
             goto cleanup;
         }
-        else if( stat( entry_name, &sb ) == -1 )
+        else
         {
```

Comment:
I'm a bit uncomfortable with the preexisting ask-for-permission (as opposed to ask-for-forgiveness) here, complete with race condition: stat then open. Granted, a directory containing certificates doesn't change often, and the use of `opendir`/`readdir` means that we don't get a consistent view of the directory if files are added, removed or renamed. But I'd still prefer to minimize such effects. In addition, the logic is getting more complex here.

If `mbedtls_x509_crt_parse_file` returns `MBEDTLS_ERR_PK_FILE_IO_ERROR`, it means that either `fopen` or `fread` failed in `mbedtls_pk_load_file`. In that case `errno` still contains the corresponding error code. Would it be possible to use this to switch to a beg-for-forgiveness approach? Always call `mbedtls_x509_crt_parse_file`, then if it returns `MBEDTLS_ERR_PK_FILE_IO_ERROR`, analyze `errno` to determine whether to count the file as silently skipped (`ENOENT`) or to error out (any other `errno` value).

This would change the behavior on non-regular, but fopen/fread-able files. For named pipes, that's rather an improvements. It would be a problem on systems where directories allow `fopen` and `fread`. What platforms allow that these days?